[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2

aorenste opened the pull request:
This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op `device_mesh::_runtime_compute_coordinate_on_dim` - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. `DeviceMesh.sym_get_coordinate` - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. `Shard._select_split_tensor` - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. `Shard.local_shard_size_and_offset` - Updated type hints to properly reflect SymInt return types.
5. Adds a config to enable compile_on_one_rank


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169552



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

aorenste committed:
WIP: [compile-one-rank] slicing

[ghstack-poisoned]

aorenste committed:
Update on "WIP: [compile-one-rank] slicing"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: [compile-one-rank] slicing"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: [compile-one-rank] slicing"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: [compile-one-rank] slicing"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: [compile-one-rank] slicing"




[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op device_mesh::_runtime_compute_coordinate_on_dim - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. DeviceMesh.sym_get_coordinate - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. Shard._select_split_tensor - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. Shard.local_shard_size_and_offset - Updated type hints to properly reflect SymInt return types.
5. Dynamo config - Always enables allow_dynamic_output_shape_ops (with TODO to reconsider if still needed).





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste commented on torch/distributed/device_mesh.py:
```diff
@@ -1106,12 +1106,27 @@ def get_coordinate(self) -> tuple[int, ...] | None:
-            # This is only valid when the current rank is part of the mesh.
-            assert self._coordinate_on_dim
-            return self._coordinate_on_dim[index]
+            if not _in_fake_mode():
```
Maybe we should put this behind a config?

ezyang commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -110,18 +110,40 @@ def _select_split_tensor(
         contiguous: bool = True,
     ) -> torch.Tensor:
         """
         Like _split_tensor() but only returns a single Tensor
```
At this point I think I would just prefer a kind of dupey comment that explains what this does lol

ezyang commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -110,18 +110,40 @@ def _select_split_tensor(
-            tensor, num_chunks, with_padding=with_padding, contiguous=contiguous
+        # We don't handle SymInt with_padding yet (because that requires extra
+        # work based on the shard)
+        if isinstance(index, int) or with_padding:
```
So this will lead to an GuardOnUnbacked error, is that right? Might be nice to have a little test that triggers this.

ezyang commented:
Is it difficult to have tests at this stage for the PR stack? I feel you now have enough kit for tests

ezyang reviewed (approved):
I don't consider the unbacked symint stuff blocking, but this is more a question for @laithsakka

aorenste commented:
> Is it difficult to have tests at this stage for the PR stack? I feel you now have enough kit for tests

It's reasonable to ask for tests at this stage. It's a little tricky because we still have the redistribute targets as non-CooR so I have to make sure the tests only rely on slicing - but I'll put something together.

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op device_mesh::_runtime_compute_coordinate_on_dim - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. DeviceMesh.sym_get_coordinate - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. Shard._select_split_tensor - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. Shard.local_shard_size_and_offset - Updated type hints to properly reflect SymInt return types.
5. Dynamo config - Always enables allow_dynamic_output_shape_ops (with TODO to reconsider if still needed).





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op device_mesh::_runtime_compute_coordinate_on_dim - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. DeviceMesh.sym_get_coordinate - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. Shard._select_split_tensor - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. Shard.local_shard_size_and_offset - Updated type hints to properly reflect SymInt return types.
5. Dynamo config - Always enables allow_dynamic_output_shape_ops (with TODO to reconsider if still needed).





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op device_mesh::_runtime_compute_coordinate_on_dim - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. DeviceMesh.sym_get_coordinate - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. Shard._select_split_tensor - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. Shard.local_shard_size_and_offset - Updated type hints to properly reflect SymInt return types.
5. Dynamo config - Always enables allow_dynamic_output_shape_ops (with TODO to reconsider if still needed).





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op device_mesh::_runtime_compute_coordinate_on_dim - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. DeviceMesh.sym_get_coordinate - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. Shard._select_split_tensor - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. Shard.local_shard_size_and_offset - Updated type hints to properly reflect SymInt return types.
5. Dynamo config - Always enables allow_dynamic_output_shape_ops (with TODO to reconsider if still needed).





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op device_mesh::_runtime_compute_coordinate_on_dim - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. DeviceMesh.sym_get_coordinate - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. Shard._select_split_tensor - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. Shard.local_shard_size_and_offset - Updated type hints to properly reflect SymInt return types.
5. Dynamo config - Always enables allow_dynamic_output_shape_ops (with TODO to reconsider if still needed).





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op `device_mesh::_runtime_compute_coordinate_on_dim` - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. `DeviceMesh.sym_get_coordinate` - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. `Shard._select_split_tensor` - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. `Shard.local_shard_size_and_offset` - Updated type hints to properly reflect SymInt return types.
5. Adds a config to enable compile_on_one_rank





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op `device_mesh::_runtime_compute_coordinate_on_dim` - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. `DeviceMesh.sym_get_coordinate` - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. `Shard._select_split_tensor` - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. `Shard.local_shard_size_and_offset` - Updated type hints to properly reflect SymInt return types.
5. Adds a config to enable compile_on_one_rank





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste committed:
Update on "[coor-slicing] Add SymInt support for DTensor mesh coordinate computation in PT2"

This change enables compile-on-one-rank for DTensor slicing by making mesh coordinate lookups symbolic-aware.

1. New custom op `device_mesh::_runtime_compute_coordinate_on_dim` - An operator that computes mesh coordinates at runtime, allowing coordinate lookups to be deferred during tracing rather than baked in as constants.
2. `DeviceMesh.sym_get_coordinate` - Now uses the custom op when in fake mode (tracing), lifting the rank map as a graph constant and deferring the actual coordinate computation to runtime.
3. `Shard._select_split_tensor` - Extended to handle SymInt indices by using torch.narrow with symbolic start/length instead of list indexing, enabling symbolic tensor partitioning.
4. `Shard.local_shard_size_and_offset` - Updated type hints to properly reflect SymInt return types.
5. Adds a config to enable compile_on_one_rank





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste commented on torch/distributed/_ops/device_mesh.py:
```diff
@@ -0,0 +1,52 @@
+    )
+
+    # We may or may not actually end up returning/using this symbol - so
+    # mark it as ignorable.
```
I made the condition when we need to do this tighter & improved the comment.
